src/Database/Query.php

```diff
@@ -889,6 +889,50 @@ public function where($conditions = null, $types = [], $overwrite = false)
         return $this;
     }
 
+    /**
+     * Adds an IN condition or set of conditions to be used in the WHERE clause for this
+     * query.
+     *
+     * This method does allow empty inputs in contrast to where() if you set
+     * $allowEmpty to true.
+     * Be careful about using it without proper sanity checks.
+     *
+     * @param string $field Field
+     * @param array $values Array of values
+     * @param bool $allowEmpty Allow empty array
+     * @return $this
+     */
+    public function whereIn($field, array $values, $allowEmpty = false)
+    {
+        if ($allowEmpty && !$values) {
+            return $this->where('1=0');
```

Comment:
We had this in the past. It resulted in a bug that deleted the whole table of a user. The bug can be triggered when you pass an empty array by mistake and the condition you are using is part of a NOT. The negation of this imposible condition results in matching all rows

Reply:
I see, what you mean. The flag makes sense to me.

Reply:
I have seen queries like
```
->where(['tag_id IN' => Hash::extract($result, '{n}.id')])
```
that suddenly broke the whole application "just because" - it was not justified and kind of goes against the ideas of the framework making certain things simpler and less buggy :)

I see the point more for the NOT IN part, but for 99% of all IN this should have never been done - and I have so many 2.x => 3.x apps broken silently along the way because of this. Not a single one exposed to the security concerns theoretically possible.

Reply:
I was once exposed to that bug myself, the application sent a lot of emails to people that should not have been on that list. It was so embarrassing and time consuming to fix the data corruption that puts on perspective that 1% of the cases.

An exception is more tolerable for me than something I cannot revert after the fact.

Reply:
Indeed - that is why it is good to have the option for both ways, defaulting to the exception case where it makes sense, and allowing the opt-in where additional checking is not necessary as nothing bad can happen here.

Reply:
I agree with you